hack/tpm2/main.go

```diff
@@ -0,0 +1,809 @@
+	// https://trustedcomputinggroup.org/resource/tcg-tpm-v2-0-provisioning-guidance
+	// https://trustedcomputinggroup.org/resource/http-trustedcomputinggroup-org-wp-content-uploads-tcg-ek-credential-profile
+	TPM2_SRK_HANDLE = tpmutil.Handle(0x81000001)
+)
+
+type TPM struct {
+	transport io.ReadWriteCloser
+}
+
+func (t *TPM) Send(input []byte) ([]byte, error) {
+	return tpmutil.RunCommandRaw(t.transport, input)
+}
+
+func (t *TPM) Close() error {
+	return t.transport.Close()
+}
+
+func seal(t transport.TPM, diskEncryptionKey []byte) (*LuksHeader, error) {
+	sealingPolicyDigest, err := calculateSealingPolicyDigest(t)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: verify the pubkey fingerprint matches the one in the pcr policy json and we have a matching digest for the pcr policy
+	// this is strictly not necessary just that unlock will fail
+
+	luks, err := createLuksHeader(t, diskEncryptionKey, sealingPolicyDigest)
+	if err != nil {
+		return nil, err
+	}
+
+	return luks, nil
+}
+
+func unseal(t transport.TPM, luks LuksHeader) ([]byte, error) {
+	tpmPub, err := tpm2.Unmarshal[tpm2.TPM2BPublic](luks.TPM2PublicBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	tpmPriv, err := tpm2.Unmarshal[tpm2.TPM2BPrivate](luks.TPM2PrivateBlob)
+	if err != nil {
+		return nil, err
+	}
+
+	srk, err := tpm2.Unmarshal[tpm2.TPM2BName](luks.TPM2SRK)
+	if err != nil {
+		return nil, err
+	}
+
+	// we need to create a primary since we don't persist the SRK
+	primary := tpm2.CreatePrimary{
+		PrimaryHandle: tpm2.TPMRHOwner,
+		InPublic:      tpm2.New2B(tpm2.ECCSRKTemplate),
+	}
+
+	createPrimaryResponse, err := primary.Execute(t)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		flush := tpm2.FlushContext{
+			FlushHandle: createPrimaryResponse.ObjectHandle,
+		}
+
+		_, flushErr := flush.Execute(t)
+		if flushErr != nil {
+			err = flushErr
+		}
+	}()
+
+	outPub, err := createPrimaryResponse.OutPublic.Contents()
+	if err != nil {
+		return nil, err
+	}
+
+	if !bytes.Equal(createPrimaryResponse.Name.Buffer, srk.Buffer) {
+		// this means the srk name does not match, possibly due to a different TPM or tpm was reset
+		// could also mean the disk was used on a different machine
+		return nil, fmt.Errorf("srk name does not match")
+	}
+
+	load := tpm2.Load{
+		ParentHandle: tpm2.NamedHandle{
+			Handle: createPrimaryResponse.ObjectHandle,
+			Name:   createPrimaryResponse.Name,
+		},
+		InPrivate: *tpmPriv,
+		InPublic:  *tpmPub,
+	}
+
+	loadResponse, err := load.Execute(t)
+	if err != nil {
+		return nil, err
+	}
+
+	policySess, policyCloseFunc, err := tpm2.PolicySession(
+		t,
+		tpm2.TPMAlgSHA256,
+		16,
+		tpm2.Salted(createPrimaryResponse.ObjectHandle, *outPub),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create policy session: %v", err)
+	}
+
+	defer policyCloseFunc()
+
+	pubKey, err := parsePCRSigningPubKey()
+	if err != nil {
+		return nil, err
+	}
+
+	loadExternal := tpm2.LoadExternal{
```

Comment:
should we switch to ecc keys for signing expected pcr values? The rsa has a limitation that `LoadExternal` only supports rsa key size upto 2048 bits, but for ecc that is not an issue. 

If we change this in future we'll have to handle both cases which could be a mess